sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -288,7 +262,15 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _currentActivity = StartActivitySourceActivity();
                 if (_currentActivity != null)
                 {
-                    _currentActivity.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);
+                    if (!(_currentActivity?.GetIsAllDataRequestedFlag() ?? true))
```

Comment:
instead of stopping and setting current activity to null (which will break context propagation we care about when activities are sampled out), can we instead either:
- add a private `IsAllDataRequested` flag to DiagScope and set it once activity is started
- or use `_currentActivity?.GetIsAllDataRequestedFlag() == true` every time

to guard any expensive operation that DiagScope does with it?

I can see just a couple of places that's needed:
- [MarkFailed](https://github.com/Azure/azure-sdk-for-net/blob/2800d6ecab44c45489a7868bfb6536fef9dc3620/sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs#L367)
- [AddTag](https://github.com/Azure/azure-sdk-for-net/blob/2800d6ecab44c45489a7868bfb6536fef9dc3620/sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs#L203)

Reply:
Just curious - is there a reason to prefer adding the IsAllDataRequested guards to each method vs setting to null? It seems like this will become easy to forget as we add new methods in the future.

Reply:
I don't have any preference, as long as we preserve Activity.Current and don't forget to close started one at the end. Agree that it might be easier to do with `_sampledOutActivity`

Reply:
assigning to new object and having an extra logic in dispose to handle it, sounds like an overhead. In case my test I can add more assertions to make sure we are not adding Tags in sampled out activities.

Reply:
More tests are always good, but I was thinking more about new methods that we may add in the future to DiagnosticScope where we will need to remember to add this check.

Reply:
if we make sure that new methods are working with sampling then, we won't miss this check.